### PR TITLE
Add additional fields to artists

### PR DESCRIPTION
**Pull request by Timo614**
Fixes: #108 

This branch adds new fields to the artists table to satisfy the requirements of issue 108.

Notes:

- It felt like over engineering to build out a new table to link back the social accounts to the artists so I kept them as optional fields on the model itself. It makes the whole process of dealing with them from the frontend much cleaner. Could be convinced otherwise though.
- We haven't added validations to date aside from the unique constraints via the db so opted to add them in this branch for the artists. I haven't modified the others as it felt out of scope but something worth us following up on after this is merged to make sure we only accept good data.
- The [validators](https://github.com/Keats/validator) have some limitations but it's something being actively worked on and the best solution I could find for the problem. You can't pass a DB connection in at the moment so there's no way to take care of unique constraints as part of the normal validation flow (which would be more ideal I feel as conflicts may occur from multiple fields).
> By default all args to a validate must be strings if you are using stable. However, if you are using nightly, you can also activate the attr_literals feature to be able to use int, float and boolean as well.
- I opted to call the `validate()` method from inside of the API vs the DB crate. Using the DB crate would limit my ability to format the errors significantly as I'd have to pass them all as one string while this other setup allows for some structure with displaying the fields with problems.
- Opted to use BadRequest in place of UnprocessableEntity as actix-web does not currently support it: https://github.com/actix/actix-web/blob/e041e9d3b722c58d757f2078ce9fed0f6211aa0d/src/httpcodes.rs#L40 -- both accomplish the same thing -- I could override it though if we have preference as I'm more used to 422 for validation errors myself although from what I found others said BadRequest was fine in this situation.
- Added auth to the artists endpoints create and update (kept index and show open)
- Removed artists destroy action (I had originally added it as part of the bn-api initial project setup but I don't think there's a need at the moment and we'll want to likely delete associated artists and perhaps even events without any other artists so it's worth doing as a separate issue if we reintroduce -- the commit history is still there if we need to resurrect this aspect)
- All of the new fields aside from bio and youtube_video_urls are optional that is somewhat optional as well you just need to provide an empty array versus no value

Let me know if the format for the error is not ideal. I mostly kept it in the output from the validator crate but could modify it if we have a specific format we want to display them in.

Examples:

Create Success

```
curl -X POST -d '{"name": "Test 1", "bio": "Biography", "website_url": "http://www.example.com", "youtube_video_urls": ["http://test.com", "http://test2.com"]}' http://localhost:8088/artists -H "Content-Type: application/json" -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiI2YTljYmU4ZS1kMDFlLTRmYWQtOGJjOC02OWE2Yjc5N2E5NjIiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM0NDYxNzU5fQ.RwQSzYsA4/TXARaskC5MMPxVIYEYTU1qhDlVhYWgZww"  -H "Accept: application/json"
{"id":"888af9d8-522d-42b8-b26a-67ea9448c3ec","name":"Test 1","bio":"Biography","website_url":"hp://www.example.com","youtube_video_urls":["http://test.com","http://test2.com"],"facebook_username":null,"instagram_username":null,"snapshat_username":null,"soundcloud_username":null,"bandcamp_username":null}
```

Create Failure

```
curl -X POST -d '{"name": "Test 2", "bio": "Biography", "website_url": "://www.example.com", "youtube_video_urls": ["http://test.com", "http://test2.com"]}' http://localhost:8088/artists -H "Content-Type: application/json" -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiI2YTljYmU4ZS1kMDFlLTRmYWQtOGJjOC02OWE2Yjc5N2E5NjIiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM0NDYxNzU5fQ.RwQSzYsA4/TXARaskC5MMPxVIYEYTU1qhDlVhYWgZww"  -H "Accept: application/json"
{"error":"Validation error","fields":{"website_url":[{"code":"url","message":null,"params":{"value":"://www.example.com"}}]}}
```

Update Success

```
curl -X PUT -d '{"name": "Test 3", "bio": "Biography", "website_url": "http://www.example.com", "youtube_video_urls": ["http://test.com", "http://test2.com"]}' http://localhost:8088/artists/888af9d8-522d-42b8-b26a-67ea9448c3ec -H "Content-Type: application/json" -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiI2YTljYmU4ZS1kMDFlLTRmYWQtOGJjOC02OWE2Yjc5N2E5NjIiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM0NDYzODgxfQ.t2FbNUbfYXoHzggdvSUI9gyS877rWcZfcaPuUHeQy0o"  -H "Accept: application/json"
{"id":"888af9d8-522d-42b8-b26a-67ea9448c3ec","name":"Test 3","bio":"Biography","website_url":"http://www.example.com","youtube_video_urls":["http://test.com","http://test2.com"],"facebook_username":null,"instagram_username":null,"snapshat_username":null,"soundcloud_username":null,"bandcamp_username":null}
```

Update Failure

```
curl -X PUT -d '{"name": "Test 3", "bio": "Biography", "website_url": "://www.example.com", "youtube_video_urls": ["http://test.com", "http://test2.com"]}' http://localhost:8088/artists/888af9d8-522d-42b8-b26a-67ea9448c3ec -H "Content-Type: application/json" -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJraWQiOm51bGwsImFsZyI6IkhTMjU2In0.eyJzdWIiOiI2YTljYmU4ZS1kMDFlLTRmYWQtOGJjOC02OWE2Yjc5N2E5NjIiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTM0NDYzODgxfQ.t2FbNUbfYXoHzggdvSUI9gyS877rWcZfcaPuUHeQy0o"  -H "Accept: application/json"
{"error":"Validation error","fields":{"website_url":[{"code":"url","message":null,"params":{"value":"://www.example.com"}}]}}
```

### CODE REVIEW

**Comment by mikethetike** [comment]
Looks really good. Will review again when it's ready

**Comment by Timo614** [comment]
Guess we can always circle back for the format changes if we decide it's worth modifying what an validation error looks like.

I rebased the branch, added the db error handling for the artists controller so it returns the conflict response when using a duplicate name, and added 404 handling to the errors logic.

There's some other errors we could handle: https://docs.diesel.rs/diesel/result/enum.Error.html but I think between the not found and the database ones we have we have a good selection and can improve upon that as we see other needs.

Tests are failing for me but also are failing on master at the moment. Something seems to be causing a deadlock with the invites logic. It looks like the other PR doesn't have that issue though post their latest changes so planning on rebasing this over master once that comes in (as I'm not sure what's triggering the deadlock condition).

File: db/src/models/artists.rs [hunk 1]
```diff
@@ -5,23 +5,53 @@ use schema::artists;
...
 
-#[derive(AsChangeset, Deserialize)]
+#[derive(AsChangeset, Deserialize, Validate)]
 #[table_name = "artists"]
-pub struct UserEditableAttributes {
+pub struct ArtistEditableAttributes {
     pub name: String,
+    pub bio: String,
+    #[validate(url)]
+    pub website_url: String,
```
[the remark this answers is not in the document]
Comment (Timo614) [reply db/src/models/artists.rs]: Oops, yup, made all those options, thanks!

**Comment by Timo614** [comment]
Sounds good -- I'm rebasing it now but will change that field to optional.

**Comment by mikethetike** [comment]
I'll create an issue for the deadlock on the invite. It seems to work on re-run, so we're not entirely sure what the cause is.

**Pull request merged**